File: server/src/middleware/strategies/tokenBucket.js
// server/src/middleware/strategies/tokenBucket.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import redisClient from '../../config/redis.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const luaScript = fs.readFileSync(
  path.join(__dirname, 'tokenBucket.lua'),
  'utf8'
);

const tokenBucketStrategy = async ({
  apiKey,
  capacity = 10,
  refillRatePerSec = 1,
}) => {
  const redisKey = `rate:${apiKey}:bucket`;
  const now = Date.now() / 1000;

  const result = await redisClient.eval(
    luaScript,
    1,               // number of KEYS
    redisKey,        // KEYS[1]
    capacity,        // ARGV[1]
    refillRatePerSec,// ARGV[2]
    now,             // ARGV[3]
    1                // ARGV[4] — tokens requested per call
  );

  const [allowed, tokensRemaining] = result;

  return {
    allowed: allowed === 1,
    remaining: Math.floor(tokensRemaining),
    resetInSeconds: Math.ceil((capacity - tokensRemaining) / refillRatePerSec),
  };
};

export default tokenBucketStrategy;

File: server/src/middleware/strategies/tokenBucket.lua
-- server/src/middleware/strategies/tokenBucket.lua
-- KEYS[1] = bucket key
-- ARGV[1] = capacity (max tokens)
-- ARGV[2] = refill rate (tokens per second)
-- ARGV[3] = current timestamp (seconds, float)
-- ARGV[4] = requested tokens (usually 1)

local bucketKey = KEYS[1]
local capacity = tonumber(ARGV[1])
local refillRate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4])

local bucket = redis.call("HMGET", bucketKey, "tokens", "lastRefillTs")
local tokens = tonumber(bucket[1])
local lastRefillTs = tonumber(bucket[2])

if tokens == nil then
  tokens = capacity
  lastRefillTs = now
end

-- Calculate how many tokens have accumulated since last refill
local elapsed = math.max(0, now - lastRefillTs)
local refillAmount = elapsed * refillRate
tokens = math.min(capacity, tokens + refillAmount)

local allowed = 0
if tokens >= requested then
  tokens = tokens - requested
  allowed = 1
end

redis.call("HMSET", bucketKey, "tokens", tokens, "lastRefillTs", now)
redis.call("EXPIRE", bucketKey, math.ceil(capacity / refillRate) * 2)

return { allowed, tokens }